### PR TITLE
🎉 Source Gitlab: add `OAuth2.0` authentication support

### LINKED ISSUE #7480: Source GitLab: support oauth
## Tell us about the problem you're trying to solve
With the release of Airbyte Cloud, we need to start supporting Oauth for this connector, since it's the recommended way of authenticating users into a SaaS application. 

If this connector doesn't support oauth already (i.e: doesn't accept a client_id and client_secret) then we need to update its spec to accept those parameters. There are two ways to do this: 

**If the connector already supports some auth mechanism like api_key**, I suggest that this be a oneof nested inside a top-level field called "authentication": 

`
{
  authentication: {
     type: object
     oneOf: [ 
       // api key,
       // oauth
     ] 
  }
}
`

**If the connector only supports webflow oauth**, then no changes are needed to the properties format and we will only need to add annotations. 

See the connector spec reference in the docs for reference on how a oneof can be implemented.

This should be done in a backwards compatible manner i.e: users currently supplying authentication info in the config's top-level should not be impacted by this change. 

## Acceptance Criteria

1. The connector supports oauth webflow authentication with client_id/client_secret in a backwards compatible manner
2. Oauth properties are annotated properly. See [this PR](https://github.com/airbytehq/airbyte/pull/6022) for an example

### PR DESCRIPTION
**Pull request by bazarnov**
## What
#7480 - Source GitLab: support oauth

## How
- added `GitlabAuthenticator` class inside of `source.py`
- extended `spec.json` with `oauthFlowSpecification`
- added `GitlabOauthFlow.java`
- added `GitlabOauthFlowTest.java` unit_test
- added `GitlabOauthFlowIntegrationTest.java` integration_test

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [X] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [X] Documentation updated 
    - [X] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [X] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [X] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/airbytehq/airbyte/7506)

### CODE REVIEW

File: airbyte-oauth/src/test/java/io/airbyte/oauth/flows/GitlabOAuthFlowTest.java
Comment (VitaliiMaltsev) [inline airbyte-oauth/src/test/java/io/airbyte/oauth/flows/GitlabOAuthFlowTest.java:79]: this test can fail in random order, because json object can be
`{"credentials":{"refresh_token":"refresh_token_response","access_token":"access_token_response"}}`
or
`{"credentials":{"access_token":"access_token_response","refresh_token":"refresh_token_response"}} and assertion will fail.
Please use `assertTrue(Maps.difference(Map.of("credentials", returnedCredentials), actualQueryParams).areEqual());` instead of
`   ` assertEquals(Jsons.serialize(Map.of("credentials", returnedCredentials)), Jsons.serialize(actualQueryParams));`

Comment (bazarnov) [reply airbyte-oauth/src/test/java/io/airbyte/oauth/flows/GitlabOAuthFlowTest.java:79]: Thanks for this suggestion! corrected.

**Review by VitaliiMaltsev — APPROVED** [review]

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py
Comment (antixar) [inline airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py:52]: are there cases when access_token is not present for this value? because you used the `get` function.

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/spec.json
Comment (antixar) [inline airbyte-integrations/connectors/source-gitlab/source_gitlab/spec.json:51]: you can use the option `const` only because `enum`, `default` and `order` will are not relevant

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py:52]: Good point, corrected.

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/spec.json
Comment (bazarnov) [reply airbyte-integrations/connectors/source-gitlab/source_gitlab/spec.json:51]: Changed.

**Review by midavadim — APPROVED** [review]

**Comment by sherifnada** [comment]
given that gitlab uses a one time refresh token, how should we move forward on this PR?

**Comment by bazarnov** [comment]
> given that gitlab uses a one time refresh token, how should we move forward on this PR? It wouldn't work in Airbyte currently until #3990 is merged. We may just want to close this and make a note in that ticket saying to reopen once it is resolved

@sherifnada 
It would be fine if we just wait for #3990 is merged.  In a meanwhile this PR is going to be blocked on #3990. WDYT?

**Comment by sherifnada** [comment]
@bazarnov sgtm

**Comment by bazarnov** [comment]
@sherifnada I believe yes, if this is critical for now, we can close it and come back afterwards

**Review by pedroslopez — COMMENTED** [review]
Note: this depends on https://github.com/airbytehq/airbyte/issues/3990 in order to work properly - let's wait until that's fully merged before releasing this connector update.

**Comment by davydov-d** [comment]
@pedroslopez got it, thanks

**Comment by davydov-d** [comment]
> Note: this depends on #3990 in order to work properly - let's wait until that's fully merged before releasing this connector update.

@YuliiaNahoha @YowanR please be informed that this PR is blocked, therefore we can not roll out a connector when it's certified until #3990 is resolved

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GitlabOAuthFlow.java
Comment (girarda) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GitlabOAuthFlow.java:103]: should we add this as a default implementation in the Interface instead of requiring all new flows to implement a deprecated method?

**Review by girarda — APPROVED** [review]
lgtm but worth mentioning this will require coordinations with TCS to release because of the spec change. @YowanR FYI

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GitlabOAuthFlow.java
Comment (girarda) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GitlabOAuthFlow.java:103]: this specific line does not require a reset, but the changes to the spec file do, yes

Comment (davydov-d) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GitlabOAuthFlow.java:103]: @girarda afaik java interfaces do not support any kind of implementation - just signatures. If you mean base class `BaseOAuth2Flow` - there are already two implementations of `completeSourceOAuth` with different signatures, one of which is deprecated. If we change it, I'm afraid it would break some of the existing oauth flow implementations.
The trick here is that deprecated APIs can not support oauth flow Implementations with dependencies on connector config (that's exactly what Gitlab implementation is) so this method is overridden here and throws an exception. Please let me know if I'm missing something.

Comment (girarda) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GitlabOAuthFlow.java:103]: we can use the [default keyword](https://stackoverflow.com/questions/31578427/what-is-the-purpose-of-the-default-keyword-in-java) to provide a default implementation.

It would look something like this
```
    default Map<String, Object> completeSourceOAuth(UUID workspaceId,
                                                    UUID sourceDefinitionId,
                                                    Map<String, Object> queryParams,
                                                    String redirectUrl,
                                                    JsonNode inputOAuthConfiguration,
                                                    OAuthConfigSpecification oauthConfigSpecification)
            throws IOException, ConfigNotFoundException, JsonValidationException {
        throw new IOException("Deprecated API. Do not use.");
    }
```

I completely agree that we shouldn't modify existing implementations. Too big of a blast radius for this PR

Comment (davydov-d) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GitlabOAuthFlow.java:103]: oh thanks for that! nevertheless it looks like we wont extract much benefit from it. This method is anyway implemented in a `BaseOAuth2Flow` so will override the default implementation, and all of the existing flow implementations (and very likely future ones) subclass `BaseOAuth2Flow`. If you insist, I will add the default implementation but in the Gitlab flow we'll have to declare it as well

Comment (girarda) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GitlabOAuthFlow.java:103]: > so will override the default implementation, and all of the existing flow implementations
fair enough. let's not do that then

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py
Comment (alafanechere) [inline airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py:59]:
```suggestion
        self.set_token_expiry_date(now, access_token_info["expires_in"])
```
this function expects two arguments: `initial_time: pendulum.DateTime, value: Union[str, int]`

**Review by alafanechere — CHANGES_REQUESTED** [review]

**Comment by alafanechere** [comment]
@davydov-d once https://github.com/airbytehq/airbyte/pull/20923 is merged would you mind trying to use the `SingleUseRefreshTokenOauth2Authenticator` directly instead of the custom authenticator you've created on top of it? It should now have the correct behavior this connector expects.

**Comment by davydov-d** [comment]
> @davydov-d once #20923 is merged would you mind trying to use the `SingleUseRefreshTokenOauth2Authenticator` directly instead of the custom authenticator you've created on top of it? It should now have the correct behavior this connector expects.

@alafanechere done!

**Comment by alafanechere** [comment]
Thanks for the changes @davydov-d !
Did you update the cloud database to set up Oauth App credentials for this connector too?

**Comment by davydov-d** [comment]
@alafanechere no yet. I was planning to do it right before publishing this PR, but since it is blocked anyway, I don't think updating DB is urgent

**Comment by alafanechere** [comment]
@davydov-d we'll manually test the connector and we'll 👍 once we get a successful sync with OAuth.
Thank you for your patience on this PR, it's an exciting change and the perfect connector for us to test the "short lived token" management we've been working on on the platform side.

**Review by alafanechere — APPROVED** [review]
I confirmed we successfully ran a sync with your latest changes 🎉 
Feel free to publish. 
The tokens in GSM might be revoked due to our tests but we're not entirely sure, so if tests that run during publish fail, it's probably because of outdated creds in GSM, feel free to refresh these or ask me to do it.

**Review by alafanechere — CHANGES_REQUESTED** [review]
Sorry, I'm actually artificially blocking the merge until we get @airbytehq/cloud-support approval too.
This new major version introduced new configuration fields which is a breaking change.
@airbytehq/cloud-support  I think you'll want to send outreach to ask users to refill their configuration after the upgrade. It will actually be an easier process now that we have OAuth in place.

**Comment by TBernstein4** [comment]
Thank you for the callout @alafanechere.  Just for clarification will there be any new fields the users will have to fill out or is private_token being replaced by OAuth completely?

I have found the workspaces that are actively using the gitlab connector and will initiate outreach soon. Would scheduling this to go out on Wednesday, January 18th at 10 AM PST be ok?

**Comment by alafanechere** [comment]
> Just for clarification will there be any new fields the users will have to fill out or is private_token being replaced by OAuth and that is it?

The private token is not used anymore.
The users that did not set a `start_date` before are now required to provide one (it was optional).
Authenticating with OAuth should work, but if they prefer to keep the API token they will need to copy it under the credentials field.

**Comment by alafanechere** [comment]
> Would scheduling this to go out on Wednesday, January 18th at 10 AM PST be ok?
Sounds good to me.
@davydov-d can you please set the OAuth config on Cloud before this date?

**Comment by TBernstein4** [comment]
Thanks for the clarification! Is there a PR for the docs to be updated with new instructions removing information about Personal Access Tokens and any other changes due to this update?

**Comment by davydov-d** [comment]
@alafanechere yes, will do

**Comment by TBernstein4** [comment]
I have sent outreach to the effected workspaces informing them of the changes and what will need to be done once the update goes out on Wednesday.

**Comment by davydov-d** [comment]
> Thanks for the clarification! Is there a PR for the docs to be updated with new instructions removing information about Personal Access Tokens and any other changes due to this update?

@TBernstein4 the [PR](https://github.com/airbytehq/airbyte/pull/20511) updating the docs with the new instructions has already been merged

**Comment by davydov-d** [comment]
@alafanechere OAuth parameters on the Cloud DB are now created

**Comment by brianjlai** [comment]
> @alafanechere OAuth parameters on the Cloud DB are now created

@davydov-d 
Due to an issue where these credentials were now be added to the configs sent from the platform to the connector, I had to remove this row from the DB. It would cause the validation to fail since the current changes to accommodate this change in `spec.json` in this PR has not been merged and deployed in the next version. Once this PR has the changed requested addressed, approved and ready to deploy, we should be able to then add the row back to the DB. But just noting here that we had to revert this.

**Comment by davydov-d** [comment]
hey @alafanechere 
I need your approval please in order to publish this change on time. I will add the OAuth config to prod DB right after I press "merge" button, otherwise it would break existing connections again

**Review by alafanechere — APPROVED** [review]
👏 
Feel free to publish, but do not merge before 10AM PST please.

**Comment by TBernstein4** [comment]
@davydov-d Is this good to merge? This roughly the time we said it would be merged

**Comment by davydov-d** [comment]
yep! :)

**Comment by TBernstein4** [comment]
Thank you!

**Comment by davydov-d** [comment]
@TBernstein4 the deployment to cloud has not been finished yet. I removed the OAuth config so the previous version should work as usual. Will restore the Oauth config as soon as the deployment succeeds

**Comment by davydov-d** [comment]
I'll make you know when the Oauth is on cloud

**Comment by TBernstein4** [comment]
Thank you for doing that @davydov-d!

**Pull request merged**
